From 2301ed809dd18ae18f085c8df12db7436850c893 Mon Sep 17 00:00:00 2001
From: Xiongfeng Wang <wangxiongfeng2@huawei.com>
Date: Fri, 31 Dec 2021 13:32:58 +0800
Subject: [PATCH] pci: do not save 'PCI_BRIDGE_CTL_BUS_RESET'
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 2301ed809dd18ae18f085c8df12db7436850c893


hulk inclusion
category: bugfix
bugzilla: 20702,https://gitee.com/openeuler/kernel/issues/I4OG3O?from=project-issue
CVE: NA
---------------------------

When I inject a PCIE Fatal error into a mellanox netdevice, 'dmesg'
shows the device is recovered successfully, but 'lspci' didn't show the
device. I checked the configuration space of the slot where the
netdevice is inserted and found out the bit 'PCI_BRIDGE_CTL_BUS_RESET'
is set. Later, I found out it is because this bit is saved in
'saved_config_space' of 'struct pci_dev' when 'pci_pm_runtime_suspend()'
is called. And 'PCI_BRIDGE_CTL_BUS_RESET' is set every time we restore
the configuration sapce.

This patch avoid saving the bit 'PCI_BRIDGE_CTL_BUS_RESET' when we save
the configuration space of a bridge.

Signed-off-by: Xiongfeng Wang <wangxiongfeng2@huawei.com>
Reviewed-by: Hanjun Guo <guohanjun@huawei.com>
Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>
Signed-off-by: Jialin Zhang <zhangjialin11@huawei.com>
Reviewed-by: Xiongfeng Wang <wangxiongfeng2@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 drivers/pci/pci.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/pci/pci.c b/drivers/pci/pci.c
index 0d7109018a91..666fcc4c37a0 100644
--- a/drivers/pci/pci.c
+++ b/drivers/pci/pci.c
@@ -1553,6 +1553,9 @@ int pci_save_state(struct pci_dev *dev)
 		pci_dbg(dev, "saving config space at offset %#x (reading %#x)\n",
 			i * 4, dev->saved_config_space[i]);
 	}
+	if (dev->hdr_type == PCI_HEADER_TYPE_BRIDGE)
+		dev->saved_config_space[PCI_BRIDGE_CONTROL / 4] &=
+			~(PCI_BRIDGE_CTL_BUS_RESET << 16);
 	dev->state_saved = true;
 
 	i = pci_save_pcie_state(dev);
-- 
2.26.2

